metaflow/plugins/kfp/kfp.py

```diff
@@ -0,0 +1,107 @@
+
+    return ordered_steps
+
+
+def run_step_op(step_name, code_url=DEFAULT_FLOW_CODE_URL):
+    """
+    Method to create a kfp container op to run a single step (here, we also execute our custom pre-start step
+    for setup as we aren't maintaining state) and then execute the given step.
+
+    TODO: This does not maintain state. The custom pre-start command used below would be removed once we have state accessible across KFP steps.
+    TODO: The public docker is a copy of the internal docker image we were using (borrowed from aip-kfp-example). Check if any stage here may need to be further modified later.
+    """
+
+    python_cmd = """ "python helloworld.py --datastore-root ", $1, " step {} --run-id ", $2, " --task-id ", $4, " --input-paths", $2"/"$5"/"$6 """.format(
+        step_name)
+    command_inside_awk = """ {{ print {0} }}""".format(python_cmd)
+    final_run_cmd = """ python helloworld.py pre-start | awk 'END{}' | sh """.format(command_inside_awk)
+
+    return dsl.ContainerOp(
+
+        name='StepRunner-{}'.format(step_name),
+        image='ssreejith3/mf_on_kfp:python-curl-git',
+        command= ['sh', '-c'],
+        arguments=[
+            'curl -o helloworld.py {}' \
+            ' && pip install git+https://github.com/zillow/metaflow.git@c722fceffa3011ecab68ce319cff98107cc49532' \
+            ' && export USERNAME="kfp-user" ' \
+            ' && {}'.format(code_url, final_run_cmd)
+            ])
+
+
+def create_flow_pipeline(ordered_steps, flow_code_url=DEFAULT_FLOW_CODE_URL, pipeline_file_path=DEFAULT_KFP_YAML_OUTPUT_PATH,
+                         return_pipeline_function=False):
+    """
+    Function that creates the KFP flow pipeline and returns the path to the YAML file containing the pipeline specification.
+    """
+
+    steps = ordered_steps
+    code_url = flow_code_url
+    print(f"\nCreating the pipeline definition to run the flow on KFP...\n")
+    print(f"\nCode URL of the flow to be converted to KFP: {flow_code_url}\n")
+
+    @dsl.pipeline(
+        name='Pipeline running MF steps',
+        description='Pipeline to experiment with MF on KFP (i.e, converting entire flow to KFP)'
+    )
+    def run_flow_pipeline():
+        """
+        This function runs the entire flow on KFP by spawning the necessary tasks
+        by invoking run_step_op for every step in the flow and handling the order of steps.
+        """
+
+        # Store the list of steps in reverse order
+        run_step_ops = [run_step_op(step, code_url) for step in reversed(steps)]
+
+        # Each step in the list can only be executed after the next step in the list, i.e., list[-1] is executed first, followed
+        # by list[-2] and so on.
+        for i in range(len(steps) - 1):
+            run_step_ops[i].after(run_step_ops[i + 1])
+
+    if return_pipeline_function:
+        return run_flow_pipeline
+
+    kfp.compiler.Compiler().compile(run_flow_pipeline, pipeline_file_path)
+    return pipeline_file_path
```

Comment:
It's cleaner that a function does one thing only, rather than either save the pipeline_file or return the pipeline_function, have this always return the pipeline_function.

The caller who wants to save the file can compile and save it themselves.  Makes for simpler code.

Reply:
Done